ziti/edge/impl/conn_test.go

```diff
@@ -0,0 +1,208 @@
+	result := make(chan error, 1)
+	result <- nil
+	return result, nil
+}
+
+func (ch *testChannel) SendWithTimeout(m *channel2.Message, timeout time.Duration) error {
+	return nil
+}
+
+func (ch *testChannel) SendPrioritizedWithTimeout(m *channel2.Message, p channel2.Priority, timeout time.Duration) error {
+	return nil
+}
+
+func (ch *testChannel) SendAndWaitWithTimeout(m *channel2.Message, timeout time.Duration) (*channel2.Message, error) {
+	panic("implement me")
+}
+
+func (ch *testChannel) SendPrioritizedAndWaitWithTimeout(m *channel2.Message, p channel2.Priority, timeout time.Duration) (*channel2.Message, error) {
+	panic("implement me")
+}
+
+func (ch *testChannel) SendAndWait(m *channel2.Message) (chan *channel2.Message, error) {
+	panic("implement me")
+}
+
+func (ch *testChannel) SendAndWaitWithPriority(m *channel2.Message, p channel2.Priority) (chan *channel2.Message, error) {
+	panic("implement me")
+}
+
+func (ch *testChannel) SendForReply(msg channel2.TypedMessage, timeout time.Duration) (*channel2.Message, error) {
+	panic("implement me")
+}
+
+func (ch *testChannel) SendForReplyAndDecode(msg channel2.TypedMessage, timeout time.Duration, result channel2.TypedMessage) error {
+	return nil
+}
+
+func (ch *testChannel) Close() error {
+	panic("implement me")
+}
+
+func (ch *testChannel) IsClosed() bool {
+	panic("implement me")
+}
+
+func (ch *testChannel) Underlay() channel2.Underlay {
+	panic("implement me")
+}
+
+func BenchmarkConnWriteBaseLine(b *testing.B) {
+	testChannel := &testChannel{}
+
+	req := require.New(b)
+
+	data := make([]byte, 1024)
+
+	b.ResetTimer()
+	for i := 0; i < b.N; i++ {
+		msg := edge.NewDataMsg(1, uint32(i), data)
+		err := testChannel.Send(msg)
+		req.NoError(err)
+	}
+}
+
+func BenchmarkConnWrite(b *testing.B) {
+	mux := edge.NewMsgMux()
+	testChannel := &testChannel{}
+	conn := &edgeConn{
+		MsgChannel: *edge.NewEdgeMsgChannel(testChannel, 1),
+		readQ:      sequencer.NewSingleWriterSeq(DefaultMaxOutOfOrderMsgs),
+		msgMux:     mux,
+		serviceId:  "test",
+	}
+
+	req := require.New(b)
+
+	req.NoError(mux.AddMsgSink(conn))
+
+	data := make([]byte, 1024)
+
+	b.ResetTimer()
+	for i := 0; i < b.N; i++ {
+		_, err := conn.Write(data)
+		req.NoError(err)
+	}
+}
+
+func BenchmarkConnRead(b *testing.B) {
+	mux := edge.NewMsgMux()
+	testChannel := &testChannel{}
+	conn := &edgeConn{
+		MsgChannel: *edge.NewEdgeMsgChannel(testChannel, 1),
+		readQ:      sequencer.NewSingleWriterSeq(DefaultMaxOutOfOrderMsgs),
+		msgMux:     mux,
+		serviceId:  "test",
+	}
+
+	var stop concurrenz.AtomicBoolean
+	defer stop.Set(true)
+
+	go func() {
+		counter := uint32(0)
+		for !stop.Get() {
+			counter += 1
+			data := make([]byte, 877)
+			msg := edge.NewDataMsg(1, counter, data)
+			mux.HandleReceive(msg, testChannel)
+		}
+	}()
+
+	req := require.New(b)
+
+	req.NoError(mux.AddMsgSink(conn))
+
+	data := make([]byte, 1024)
+
+	b.ResetTimer()
```

Comment:
Could be wrong, but I believe the _best_ way to write a golang benchmark is to engineer the actual `Benchmark` function, so that this stuff is done in an `init` or some other setup infrastructure, because even though this stuff is outside the `i < b.N` loop, it'll still affect the results a little bit.

Reply:
the docs here: https://golang.org/pkg/testing/ point to `b.ResetTimer()` as a way to avoid the issue.